### PR TITLE
DOC: Add column name metadata to spec

### PR DESCRIPTION
**Pull request by cpcloud**
This adds an additional metadata field to the spec to allow faithful
reproduction of names of column indexes.

cc @wesm @martindurant @jreback

### CODE REVIEW

**Comment by cpcloud** [comment]
xref: https://issues.apache.org/jira/browse/ARROW-1586

File: doc/source/developer.rst [hunk 1]
```diff
@@ -45,6 +45,7 @@ So that a ``pandas.DataFrame`` can be faithfully reconstructed, we store a
 .. code-block:: text
 
    {'index_columns': ['__index_level_0__', '__index_level_1__', ...],
+    'column_index_names': [<column index level name 0>, <column index level name 1>, ...],
```
Comment (jreback) [inline doc/source/developer.rst]: should also store: ``columns_index_dtypes`` to faithfully restore the dtypes as well

Comment (jreback) [reply doc/source/developer.rst]: this one to?

**Comment by jreback** [comment]
thanks @cpcloud

File: doc/source/developer.rst [hunk 2]
```diff
@@ -106,6 +108,8 @@ As an example of fully-formed metadata:
 .. code-block:: text
 
    {'index_columns': ['__index_level_0__'],
+    'column_index_names': [None],
+    'column_index_dtypes': ['object'],
```
Comment (wesm) [inline doc/source/developer.rst]: We may want to enrich these types to include both the pandas type and the numpy type

Comment (cpcloud) [reply doc/source/developer.rst]: Yep, I'm going to follow up here.

**Pull request merged**